Leetcode/aoc_2015_d4.py

```diff
@@ -0,0 +1,140 @@
+Then, it's as simple as looping until it finds the first hash that 
+has 5 leading zeroes. But since the number that is combined with 
+the secret key can't have a leading zero itself, we can just use a count
+variable that goes each loop.
+Combine the secret key with the count variable with string concate, find
+the current hash of that, and if it doesn't lead with 5 zeroes, loop again.
+
+I also didn't understand the problem at first. I was confused as to how the
+answer was found, and what it was doing to get it. I had looked up an online
+string to MD5 converter and played with that for a bit. Putting in the strings
+that the problem gave me until I put in the right combination and got the answer
+that it was looking for.
+
+So looking it up, Python has a built-in called hashlib that can
+convert a string into an MD5 hash.
+'''
+import hashlib
+import time # used to get execution time.
+
+with open("aoc_2015_d4_input.txt") as file:
+	h = file.read()
+
+
+def find_hash(h: str) -> int:
+	# set target hash to 0 for now. placeholder so that the
+	# while loop can run once. Could compensate by doing a 
+	# do while instead of this.
+	target_hash = "0"
+	# the iterable number that will get added to the string.
+	# starts at 0 but gets added right away in the while loop
+	# so that it truly starts at 1.
+	find_hash = 0
+	# slice off just the first five characters to check if it's
+	# leading with 5 zeroes.
+	while target_hash[:5] != "00000":
+		# Starts at 1 since the answer can't start with a leading
+		# zero. And by adding the number first, we then perform the
+		# calculations, so that once the next loop starts, if the
+		# previous hash started with 5 zeroes, the next loop wouldn't
+		# run, thereby not adding one to the find_hash.
+		find_hash += 1
+		# concate the secret key and the iterable then encodes it
+		# to be turned into an md5 hash.
+		str2hash = (h + str(find_hash)).encode()
+		# gets the md5 hash, then converts it into a readable string.
+		target_hash = hashlib.md5(str2hash).hexdigest()
+
+	return find_hash
+
+'''
+ ----- Thoughts on part 1 after completing it. -----
+This one was a bit tricky. I understood it conceptually, but I didn't know
+how to put it in code as easily since I didn't know how the hashlib and md5
+worked. It took a bit of trial and error, but understanding it conceptually
+made coding it a bit easier, I think.
+Since it must go through numbers by iteration, it can take a good amount of time
+find the right string that leads with 5 zeroes.
+If I really wanted to improvet this, I could probably do something similar to my
+java mega millions program and have it multithreaded with each loop going through
+say 999_999 numbers. e.x. thread 1 does 1 to 100_000, thread 2 does 1_000_000 to
+1_999_999 numbers. But that's micro-optimization, since there's no guarantee it'd
+be a number that high. For example, my secret key had an answer that was quite low.
```

Comment:
This does work, it's essentially how Bitcoin miners speed up to utilize all of the available hardware resources.

Your analysis is half right, and you're right that if you're looking for the _lowest_ one then there's no better way than to start from zero and work your way up - because even if you did multithread and one of your threads found a valid hash, you would have no way of knowing if it's the lowest valid one without checking everything that came beforehand.

In the context of Bitcoin mining optimization and parallelization, this does work, since Bitcoin just requires _any_ valid hash, not the lowest one. And if we're talking about one that has 5 zeros, there's actually a lot of them because all the valid ones are in the range of `00000000000000000000000000000000` to `00000fffffffffffffffffffffffffff`. So `00000000000000000000000000000001` is valid, `00000000000000000000000000000002` is valid, and so on.

![image](https://github.com/ToekneeT/interview-prep/assets/2053862/71253235-a30d-4b5f-87af-54687ef0e50b)
 
ChatGPT says $2^{108}$ hashes. And since one of the properties of hashes is that we can't predict what the resulting hash looks like (`md5(1)` looks completely different to `md5(2)`, with no pattern), nonce=1 is equally likely to be 5-zero valid as nonce=1000, or nonce=10000, and so on.

That is to say that multithreading to search different nonce ranges is very valid, and it will be a huge speedup.

Reply:
That is a lot of hashes. Glad to know my analysis is somewhat right.